File: Substring_with_concatenation_of_all_words/Substring_with_concatenation_of_all_words.py
def substring_with_concatenation_of_all_words(s, words):

    def generate_permutations(word, current_perm=[]):
        # Base case: If no words left, add the current permutation to the result

        if not word:
            return [''.join(current_perm)]

        result = []
        for j in range(len(word)):
            # Choose the current word and recursively generate permutations
            remaining = word[:j] + word[j + 1:]  # Remove the chosen word
            result += generate_permutations(remaining, current_perm + [word[j]])

        return result

    output = []

    all_concatenations = generate_permutations(words)

    for concatenation in all_concatenations:
        for i in range(len(s)):
            if s[i:i+len(concatenation)] == concatenation:
                output.append(i)

    print(output)

substring_with_concatenation_of_all_words("barfoothefoobarman", ["foo", "bar"])
substring_with_concatenation_of_all_words("wordgoodgoodgoodbestword", ["word","good","best","word"])
substring_with_concatenation_of_all_words("barfoofoobarthefoobarman",["bar", "foo", "the"])

File: Substring_with_concatenation_of_all_words/Substring_with_concatenation_of_all_words.java
import java.util.ArrayList;
import java.util.List;

public class Main {

    public static List<String> generatePermutations(List<String> word, List<String> current_perm) {
        List<String> result = new ArrayList<>();

        if (word.isEmpty()) {
            result.add(String.join("", current_perm));
            return result;
        }

        for (int i = 0; i < word.size(); i++) {
            String w = word.get(i);

            List<String> remaining = new ArrayList<>(word);
            remaining.remove(i);

            List<String> newPerm = new ArrayList<>(current_perm);
            newPerm.add(w);

            result.addAll(generatePermutations(remaining, newPerm));
        }

        return result;

    }

    public static List<Integer> findSubstring(String s, List<String> words) {
        List<Integer> output = new ArrayList<>();

        List<String> all_concatenations = generatePermutations(words, new ArrayList<>());

        for (String concatenation : all_concatenations) {
            int len = concatenation.length();

            for (int j = 0; j < s.length() - len; j++) {
                if (s.substring(j, j+len).equals(concatenation)) {
                    output.add(j);
                }
            }
        }
        return output;
    }

    public static void main(String[] args) {
        System.out.println(findSubstring("barfoothefoobarman", List.of("foo", "bar")));
        System.out.println(findSubstring("wordgoodgoodgoodbestword", List.of("word", "good", "best", "word")));
        System.out.println(findSubstring("barfoofoobarthefoobarman", List.of("bar", "foo", "the")));


    }

}